invokeai/backend/stable_diffusion/extensions_manager.py

```diff
@@ -0,0 +1,167 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from contextlib import ExitStack, contextmanager
+from functools import partial
+from typing import TYPE_CHECKING, Callable, Dict
+
+import torch
+from diffusers import UNet2DConditionModel
+
+from invokeai.backend.util.devices import TorchDevice
+
+if TYPE_CHECKING:
+    from invokeai.backend.stable_diffusion.denoise_context import DenoiseContext
+    from invokeai.backend.stable_diffusion.extensions import ExtensionBase
+
+
+class ExtCallbacksApi(ABC):
+    @abstractmethod
+    def setup(self, ctx: DenoiseContext, ext_manager: ExtensionsManager):
+        pass
+
+    @abstractmethod
+    def pre_denoise_loop(self, ctx: DenoiseContext, ext_manager: ExtensionsManager):
+        pass
+
+    @abstractmethod
+    def post_denoise_loop(self, ctx: DenoiseContext, ext_manager: ExtensionsManager):
+        pass
+
+    @abstractmethod
+    def pre_step(self, ctx: DenoiseContext, ext_manager: ExtensionsManager):
+        pass
+
+    @abstractmethod
+    def post_step(self, ctx: DenoiseContext, ext_manager: ExtensionsManager):
+        pass
+
+    @abstractmethod
+    def pre_unet(self, ctx: DenoiseContext, ext_manager: ExtensionsManager):
+        pass
+
+    @abstractmethod
+    def post_unet(self, ctx: DenoiseContext, ext_manager: ExtensionsManager):
+        pass
+
+    @abstractmethod
+    def post_apply_cfg(self, ctx: DenoiseContext, ext_manager: ExtensionsManager):
+        pass
+
+
+class ProxyCallsClass:
+    def __init__(self, handler):
+        self._handler = handler
+
+    def __getattr__(self, item):
+        return partial(self._handler, item)
+
+
+class CallbackInjectionPoint:
+    def __init__(self):
+        self.handlers = {}
+
+    def add(self, func: Callable, order: int):
+        if order not in self.handlers:
+            self.handlers[order] = []
+        self.handlers[order].append(func)
+
+    def __call__(self, *args, **kwargs):
+        for order in sorted(self.handlers.keys(), reverse=True):
+            for handler in self.handlers[order]:
+                handler(*args, **kwargs)
+
+
+class ExtensionsManager:
```

Comment:
I think your suggestions miss one initial intention of this implementation - that extensions can themselves define and call callbacks, so that's mean:
1) Enum should be defined as `(str, Enum)` so that extensions can define custom ones with strings
2) To be able call this custom callbacks, extension should have reference to extension manager, so I added it to all args with this idea in mind
3) args/kwargs helps in this case for custom callbacks to work, as without it, only remaining option is to pass all info in `ctx.extra`
Also about `CallbackInjectionPoint` I too initially wanted to make it simple as your code, but then decided to have more strict order, so that by my code I ensure that if we have multiple callbacks with same `order`, then they will be called exactly in order of registration.
About tests - I suck at it, you are welcome to create them :D

Reply:
> I think your suggestions miss one initial intention of this implementation - that extensions can themselves define and call callbacks,

This is the first time I'm hearing about this. This feels out of scope right now. It would probably warrant its own design doc (or significant modification to the current one). Is there  any near-term use case for this?

> Also about CallbackInjectionPoint I too initially wanted to make it simple as your code, but then decided to have more strict order, so that by my code I ensure that if we have multiple callbacks with same order, then they will be called exactly in order of registration.

I *think* the existing code might already do this, but if it's a requirement then we should definitely add a unit test for it. If it's not already working, then it should be easy to achieve: the [sorted()](https://docs.python.org/3/library/functions.html#sorted) function is guaranteed to be stable (i.e. it preserves the order of elements that compare equal).

Reply:
> This is the first time I'm hearing about this. This feels out of scope right now. It would probably warrant its own design doc (or significant modification to the current one). Is there any near-term use case for this?

There was and idea, that tiled node can call `tiled_denoise_setup` in original `setup`, and nodes whose actions changes in tiled generation, will be able to listen for this callback and make changes according to it.
For example this can be used by t2i adapter, because in normal generation we prefer run it once before generation and save adapter states, while in tiled generation we will required to run model on each tile.

> I think the existing code might already do this, but if it's a requirement then we should definitely add a unit test for it. If it's not already working, then it should be easy to achieve: the [sorted()](https://docs.python.org/3/library/functions.html#sorted) function is guaranteed to be stable (i.e. it preserves the order of elements that compare equal).

Probably bugs from being writing code with temperature :D But I think it should work same, yes.

Reply:
It's a bit different interaction type and I can't say it's better/worse because it's simply different. There some moments which better can be done with such custom callbacks and some with extra variables in context.
For example about tiled generation:
 - callbacks badly suits here for passing information about current tile IMHO, but extra variable fits it
 - but if we want to reinitialize t2i for tiled generation I feel that callbacks suits more, thought it possible to do with extra fields
But I also wants to hear @dunkeroni opinion.

About compatibility, it's almost completely compatible code, except one moment - extension manager as second argument to callbacks.
Because both - changing enum type and *args, **kwargs can be done silently for all logic,

Reply:
> I don't think I'm fully understanding what you've described. Why is this better than setting a flag in the DenoiseContext to affect how other nodes behave?
> 
> It feels like it adds some complexity to the ExtensionsManager, but the value isn't clear to me.
> 
> Also, is this something that would be easy to add later if it's needed? Or do you think it would be a major change later?

Yes, it causes some more complexity, and no we don't know the use of it right now; but we do know that there is potential value in it. Most of that value is in the fact that we don't know how the pipeline might get extended in the future. Whether that value is worth the complexity is certainly debatable, and I have no qualms with taking it out for the sake of a more straightforward API for the initial release. I will do my best to illustrate why it is worth considering, and I welcome your opinions to the contrary:

Primarily it makes things more flexible in the context of interoperability of extensions. Tiling is the method of generation subsampling that Invoke currently officially supports (I will also point out that support is extremely recent), but it is not the _only_ subsampling method used in diffusion. Scalecrafter and Demofusion both use forms of dilation subsampling to maintain the full image and prompt context in a lower resolution while computing. If an extension like Controlnet was only compatible with tiling by checking for flag values and adjusting accordingly, then the only way to make a custom version compatible with other forms of subsampling would be to add another flag and rewrite the existing code around also checking for dilation. Note that DemoFusion uses both multidiffusion tiling and dilation in the same pipeline, so an extension or set of extensions that can react to both might not be able to distinguish the correct subsample. The extensions that trigger the changes would have to immediately delete the extra flags after they are used to avoid lingering flags. 
It would be better in that case to have a specialized callback for tiling and a different one for dilation so that the extension receives an explicit call each (sub)step and knows what changes to make. Then since dilation is not supported in the official extension, something could inherit from controlnet and only add the dilation handling callback while still wholly utilizing the official upstream code to remain compatible with future changes. 

The other, less concrete benefit is that we do not know that our current set of injection points are the best for all future pipelines. We have defined a set of override and modification points based on what we believe is the typical process a diffusion model should follow. However, future adapter networks or new model architectures might introduce new critical points into that pipeline, and they might be for models that Invoke does not want to officially support right away but could do so indirectly through extensions.

Reply:
Thanks for this explanation. To make sure we're on the same page, here's some pseudo-code for the situation you've described. This is what it would look like if we **do not** allow extensions to make calls to the extension manager (with comments highlighting how this results in some code duplication):

```python
class MyCustomControlnetExt(ControlNetExt):

    @callback(ExtensionCallbackType.PRE_UNET)
    def pre_unet_callback(self, ctx: DenoiseContext):
    	if "dilation_config" in ctx.extras:
    	    self._run_controlnet_with_dilation(...)

    	# The code below this point would have to be copy-pasted from ControlNetExt,
    	# it can't just be inherited!!!
    	# -------------------------------------------
        elif ctx.multidiffusion_tiles is not None:
            self._run_controlnet_with_multidiffusion(ctx)
        else:
            self._run_controlnet(...)

    ...
```

Let me know if I got that wrong.

The code duplication isn't ideal, but it feels workable enough that I'd still lean towards the simplicity of not allowing extensions to call the extension manager (for now).

---

It seems like there are two decisions to be made:

1. Should the ExtensionManager be generalized to allow any callback (represented by a string)? Or just a defined set of callbacks?
- If we generalize it, a node author could use it to write a new 'primary' node with an entirely new set of callbacks that supports extensions.
- In that case, we'd want to make the ExtensionManager more abstract - it shouldn't have methods like patch_unet(...), since it might be used for workflows that don't use a UNet.

2. Should extensions be allowed to call the ExtensionManager? i.e. callbacks within callbacks
- This is mainly what was being discussed in the example above.


My answer to both of these questions is: "not right now". Let's keep it simple for the first version. It's much easier to add complexity and features in the future than to remove them.

(Let me know if you disagree.)

Reply:
In that case, you could still inherit the upstream compatibility code by using `super().pre_unet_callback(ctx)` after the comment. It is just relying on some extra handling to make sure that `ctx.extras.dilation_config` is deleted whenever `pre_unet_callback` is supposed to have a tile instead. Otherwise CNet would incorrectly default to the last used dilation. But that case is easy enough to handle for anyone deep enough in the weeds to need it, so until it becomes a problem we don't need to design around it.

**A1**: "Not right now" seems fine. The types of workflow custom callbacks enable are often complicated to the point that the only real user is the person who wrote it, or big enough that it would also require rewriting other parts of this API. Restrictions on available callback points also let us more clearly document how things should be done instead of just what can be done, which improves alignment with multiple custom extensions.

**A2**: It seems reasonable to say that callbacks should not trigger other callbacks; with the exception of transfer functions, which we are not implementing for now because they are rife with potential user mistakes and difficult to explain. A transfer function would be any extension that returns a combination of other extension results (to transition from one to another mid-process, e.g. DemoFusion cosine decay between dilation and multidiffusion). In that case, the transfer function extension would have its own internal ExtensionManager(s) that it can call to get results from its sub-pipelines.

So a callback should not have visibility of the other extensions, except for the data they have already put into the DenoiseContext. Overrides that replace structural levels of the code which contain callbacks are the only case where an extension needs to interact with its own ExtensionManager.